SimGeneral/CaloAnalysis/plugins/MtdTruthAccumulator.cc

```diff
@@ -0,0 +1,937 @@
+    IfLogDebug(DEBUG, messageCategoryGraph_)
+        << std::fixed << std::setprecision(3) << "SimCluster from CP with:"
+        << "\n  charge " << sc.charge() << "\n  pdgId  " << sc.pdgId() << "\n  energy " << sc.energy()
+        << " GeV\n  eta    " << sc.eta() << "\n  phi    " << sc.phi()
+        << "\n  number of cells = " << sc.hits_and_fractions().size() << std::endl;
+    for (unsigned int i = 0; i < sc.hits_and_fractions().size(); ++i) {
+      DetId id(sc.hits_and_fractions()[i].first);
+      IfLogDebug(DEBUG, messageCategoryGraph_)
+          << std::fixed << std::setprecision(3) << " hit " << sc.hits_and_fractions()[i].first << " on layer "
+          << geomTools_.getLayer(id) << " at time " << sc.hits_and_times()[i].second << " ns" << std::endl;
+    }
+    IfLogDebug(DEBUG, messageCategoryGraph_) << "--------------\n";
+  }
+  IfLogDebug(DEBUG, messageCategoryGraph_) << std::endl;
+#endif
+
+  // save the SimCluster orphan handle so we can fill the calo particles
+  auto scHandle = event.put(std::move(output_.pSimClusters), "MergedMtdTruth");
+
+  // reserve for the best case scenario: already splitted
+  output_.pMtdSimLayerClusters->reserve(scHandle->size());
+  output_.pMtdSimTracksters->reserve(scHandle->size());
+
+  uint32_t SC_index = 0;
+  uint32_t LC_index = 0;
+  for (const auto &sc : *scHandle) {
+    auto const &hAndF = sc.hits_and_fractions();
+    auto const &hAndE = sc.hits_and_energies();
+    auto const &hAndT = sc.hits_and_times();
+    // create a vector with the indexes of the hits in the simCluster
+    std::vector<int> indexes(hAndF.size());
+    std::iota(indexes.begin(), indexes.end(), 0);
+    // sort the hits indexes based on subdetector, layer, module, then pixel row and column
+    std::sort(indexes.begin(), indexes.end(), [&](int a, int b) {
+      DetId aa = hAndF[a].first;
+      DetId bb = hAndF[b].first;
+      if (geomTools_.isETL(aa) != geomTools_.isETL(bb))
+        return (int)geomTools_.isETL(aa) < (int)geomTools_.isETL(bb);
+      else if (geomTools_.isETL(aa) and (geomTools_.getLayer(aa) != geomTools_.getLayer(bb)))
+        return geomTools_.getLayer(aa) < geomTools_.getLayer(bb);
+      else if (geomTools_.getModule(aa) != geomTools_.getModule(bb))
+        return geomTools_.getModule(aa) < geomTools_.getModule(aa);
+      else if (geomTools_.getPixelInModule(aa, m_detIdToPos[aa]).second ==
+                   (int)geomTools_.getPixelInModule(bb, m_detIdToPos[bb]).second and
+               std::abs(geomTools_.getPixelInModule(aa, m_detIdToPos[aa]).first -
+                        (int)geomTools_.getPixelInModule(bb, m_detIdToPos[bb]).first) < 2)
+        return geomTools_.getPixelInModule(aa, m_detIdToPos[aa]).first <
+               (int)geomTools_.getPixelInModule(bb, m_detIdToPos[bb]).first;
+      else
+        return geomTools_.getPixelInModule(aa, m_detIdToPos[aa]).second <
+               (int)geomTools_.getPixelInModule(bb, m_detIdToPos[bb]).second;
+    });
```

Comment:
@AuroraPerego as discussed, this works ok for BTL. In ETL anyway DetId is not enough to identify the hit location, row and column are also needed. And this lexicographic ordering does not necessarily put always close to each other adjacent hits, in case of very high occupancy, do you agree?